### PR TITLE
value packing

### PR DESCRIPTION
**Pull request by tj-sun**
This PR fixes bugs in buffer size calculation when packing values instead of packing pointers.

### CODE REVIEW

File: pyop2/base.py [hunk 1]
```diff
@@ -4327,7 +4327,7 @@ def build_itspace(args, iterset):
             if arg._uses_itspace:
                 _block_shape = arg._block_shape
                 if block_shape and block_shape != _block_shape:
-                    raise IndexValueError("Mismatching iteration space size for argument %d" % i)
+                    pass  # Allow different block shape
```
Comment (miklos1) [inline pyop2/base.py]: So what is this thing with the mismatching block size?

File: pyop2/sequential.py [hunk 2]
```diff
@@ -1096,20 +1096,20 @@ def extrusion_loop():
...
             # Readjust size to take into account the size of a vector space
-            _dat_size = (arg.data.cdim,)
-            _buf_size = [sum([e*d for e, d in zip(_buf_size, _dat_size)])]
-            _loop_size = [_buf_size[i]//_dat_size[i] for i in range(len(_buf_size))]
+            _dat_size = [_arg.data.cdim for _arg in arg]
+            _buf_size = [sum([e*d for e, d in zip(_loop_size, _dat_size)])]
         else:
             _dat_size = arg.data.dims[0][0]  # TODO: [0][0] ?
-            _buf_size = [e*d for e, d in zip(_buf_size, _dat_size)]
+            _buf_size = [e*d for e, d in zip(_loop_size, _dat_size)]
```
Comment (miklos1) [inline pyop2/sequential.py]: Can you explain what is what here, and why does `_buf_size` change to `_loop_size`?

File: pyop2/sequential.py [hunk 3]
```diff
@@ -1096,20 +1096,20 @@ def extrusion_loop():
...
         _buf_decl[arg] = arg.c_buffer_decl(_buf_size, count, _buf_name[arg], is_facet=is_facet)
         _tmp_decl[arg] = arg.c_buffer_decl(_buf_size, count, _tmp_name[arg], is_facet=is_facet,
                                            init=False)
+        facet_mult = 2 if is_facet else 1
         if arg.access not in [WRITE, INC]:
-            _itspace_loops = '\n'.join(['  ' * n + itspace_loop(n, e) for n, e in enumerate(_loop_size)])
+            _itspace_loops = '\n'.join(['  ' * n + itspace_loop(n, e*facet_mult) for n, e in enumerate(_loop_size)])
```
Comment (miklos1) [inline pyop2/sequential.py]: Why did this work before?

File: pyop2/sequential.py (hunk 2 above)
Comment (tj-sun) [reply pyop2/sequential.py]: The buffer sizes seem wrong form many cases, e.g. `tests/regression/test_2dcohomology.py::test_betti1[space5]`

The generated wrapper is
```C
double buffer_arg0_0[20] __attribute__((aligned(16))) = {0.0};
double buffer_arg1_0[40] __attribute__((aligned(16)));
double buffer_arg2_0[20] __attribute__((aligned(16)));
double buffer_arg3_0[40] __attribute__((aligned(16)));
```
whereas the correct sizes are
```C
double buffer_arg0_0[20] __attribute__((aligned(16))) = {0.0};
double buffer_arg1_0[6] __attribute__((aligned(16)));
double buffer_arg2_0[20] __attribute__((aligned(16)));
double buffer_arg3_0[30] __attribute__((aligned(16)));
```
As far as I see, it seems to be a consequence of computing the buffer sizes from iteration extents.

File: pyop2/sequential.py (hunk 3 above)
Comment (tj-sun) [reply pyop2/sequential.py]: I think this was taken care of by `map_init` which has a flag `is_facet`

https://github.com/OP2/PyOP2/blob/master/pyop2/sequential.py#L1071

**Comment by tj-sun** [comment]
This commit passes all tests except fusion.

The [`iterset_loop_body`](https://github.com/OP2/PyOP2/blob/value-packing-pr/pyop2/sequential.py#L1123) is a bit ugly as it was previously modeled on top of `IterationSpace`, but I think it handles mixed forms correctly now.

File: pyop2/sequential.py [hunk 4]
```diff
@@ -843,13 +843,12 @@ def compile(self):
         # Blow away everything we don't need any more
         del self._args
         del self._kernel
-        del self._itspace
```
Comment (miklos1) [inline pyop2/sequential.py]: You sure don't want to delete `self._iterset` instead?

**Comment by miklos1** [comment]
Why does this change break fusion?

**Comment by tj-sun** [comment]
It's just that I haven't updated the fusion code since removing `IterationSpace`, there are a few dependencies.

File: pyop2/sequential.py (hunk 4 above)
Comment (tj-sun) [reply pyop2/sequential.py]: Make sense. Amended.

**Comment by miklos1** [comment]
> It's just that I haven't updated the fusion code since removing IterationSpace, there are a few dependencies.

Would that be as mechanical as the simple updates in `sequential.py`, or is there anything more challenging? I have an aversion to deliberately leaving code broken...

File: test/unit/test_indirect_loop.py [hunk 5]
```diff
@@ -124,6 +124,7 @@ def test_mismatching_indset(self, iterset, x):
             op2.par_loop(op2.Kernel("", "dummy"), iterset,
                          x(op2.WRITE, op2.Map(iterset, op2.Set(nelems), 1)))
 
+    @pytest.mark.skip("allow different block size")
```
Comment (miklos1) [inline test/unit/test_indirect_loop.py]: Just remove this test, it does not apply anymore.

**Comment by tj-sun** [comment]
I mainly needs to modify the fusion code, hopefully nothing needs to be done in `sequential.py`, so shouldn't be too bad (I think).

File: test/unit/test_indirect_loop.py (hunk 5 above)
Comment (tj-sun) [reply test/unit/test_indirect_loop.py]: Cool.

**Comment by miklos1** [comment]
> I mainly needs to modify the fusion code, hopefully nothing needs to be done in sequential.py, so shouldn't be too bad (I think).

OK, either that, or if we officially don't care about the fusion code, then file a separate PR that extirpates all of it.

File: pyop2/base.py [hunk 6]
```diff
@@ -4272,39 +4149,18 @@ def iteration_region(self):
...
-def build_itspace(args, iterset):
-    """Creates an class:`IterationSpace` for the :class:`ParLoop` from the
-    given iteration set.
-
-    Also checks that the iteration set of the :class:`ParLoop` matches the
+def check_iterset(args, iterset):
+    """Checks that the iteration set of the :class:`ParLoop` matches the
     iteration set of all its arguments. A :class:`MapValueError` is raised
     if this condition is not met.
-
```
Comment (miklos1) [inline pyop2/base.py]: This looks like the accidental removal of a healthy blank line (?)

File: pyop2/sequential.py [hunk 7]
```diff
@@ -1132,45 +1131,50 @@ def itset_loop_body(i, j, shape, offsets, is_facet=False):
...
             elif arg._is_mat:
                 continue
             elif arg._is_dat:
-                loop_size = shape[0]*mult
-                _itspace_loops, _itspace_loop_close = itspace_loop(0, loop_size), '}'
-                _scatter_stmts = arg.c_buffer_scatter_vec(count, i, j, offsets, _buf_name[arg])
-                _buf_offset, _buf_offset_decl = '', ''
+                arg_scatter = []
+                offset = 0
+                for i, m in enumerate(arg.map):
```
Comment (miklos1) [inline pyop2/sequential.py]: Is this loop to support mixed?

File: pyop2/base.py (hunk 6 above)
Comment (tj-sun) [reply pyop2/base.py]: Think so. Added back.

File: pyop2/sequential.py (hunk 7 above)
Comment (tj-sun) [reply pyop2/sequential.py]: Yes indeed.

**Pull request merged**